src/viam/rpc/dial.py

```diff
@@ -222,7 +223,7 @@ async def dial(address: str, options: Optional[DialOptions] = None) -> ViamChann
 
     runtime = _Runtime()
 
-    path, path_ptr = runtime.dial(address, opts)
+    path, path_ptr = await asyncio.to_thread(runtime.dial, address, opts)
```

Comment:
I believe the `runtime.dial` function is not thread safe due to the reverse semaphore not being thread safe. We could make the semaphore (`PointerCounter`) thread safe by adding locks, or we could run the actual blocking part of the function (`_lib.dial`, L191) in `asyncio.to_thread`

Reply:
I think we would want both, no? Adding locks on the semaphore would fix other thread safety issues that have been cropping up like the one with `_ptr` I posted in slack, while `asyncio.to_thread` will fix the blocking issue.

Reply:
Should the lock go on the `_Runtime` object to be used next to the semaphore, or should we add to `PointerCounter` itself?

Reply:
I think the lock should go on the PointerCounter object. Something like

```python3
class PointerCounter:
    def __init__(self) -> None:
        self._event = Event()
        self._count = 0
        self._event.set()
        self._lock = Lock() <<<<<<<<

    def increment(self) -> int:
        with self._lock: <<<<<<<<
            self._count += 1
            self._event.clear()
            return self._count

    def decrement(self) -> int:
        with self._lock: <<<<<<<<
            assert self._count > 0, "Pointer count cannot go below zero"
            self._count -= 1
            if self._count == 0:
                self._event.set()
            return self._count
    
    ...
```

Reply:
Yah that's about what I prototyped.

That also means that increment and decrement need to be `async` though

Reply:
The bubbling up of `async` on `PointerCounter` goes all the way to changing `ViamChannel`, and I think also does not fully solve the problem.

I think instead we want a lock on `_Runtime`, as without one, multiple threads opening a connection simultaneously can both hit `if not hasattr(cls, "_shared"):` and then do all sorts of bad things

Reply:
Should we not lock both `_Runtime` and `PointerCounter`?

Reply:
I'm okay with this.

Are you okay with me using a threading.Lock on PointerCounter (like with _Runtime) rather than asyncio.Lock, for thread safety?

Reply:
Also to avoid the `async` proliferation that will happen with an asyncio Lock

Reply:
Yeah if the reason for doing this is thread safety then let's use a threading lock rather than an asyncio one

Reply:
Updated